File: Python/42.py
# Notes
    # This seems like the valid subarray problem where it's likely easier to generate solution
        # index by index instead of trying to find a solution using subarrays
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) == 1: return 0
        
        # Generate left maxes
        left_maxes = [0]
        for i in range(len(height) - 1):
            if height[i] > left_maxes[-1]:
                left_maxes.append(height[i])
            else:
                left_maxes.append(left_maxes[-1])

        # Generate right maxes
        right_maxes = [0]
        for i in range(len(height) - 1, 0, -1):
            if height[i] > right_maxes[-1]:
                right_maxes.append(height[i])
            else:
                right_maxes.append(right_maxes[-1])
        right_maxes = right_maxes[::-1]

        # At each index, return the minimum of the left/right maxes minus the current height
        total = 0
        for i, h in enumerate(height):
            if right_maxes[i] and left_maxes[i]:
                total += max(0, min(left_maxes[i], right_maxes[i]) - height[i])
        return total

File: C++/42.cpp
/*
- As usual for hard array problems, think of it index by index
- Water contained at each index equals the minimum of it's left/right maxes minus its height
- Remember to take advantage of max/min function to avoid conditional branching
*/
class Solution {
public:
    int trap(vector<int>& height) {
        vector<int> left_maxes = {0};
        vector<int> right_maxes = {0};

        int cur = 0;
        int lm = 0;
        for(int i = 1; i < height.size(); i++){
            cur = height[i - 1];
            lm = max(cur, lm);
            left_maxes.push_back(lm);
        }

        cur = 0;
        int rm = 0;
        for(int i = height.size() - 2; i >= 0; i--){
            cur = height[i + 1];
            rm = max(cur, rm);
            right_maxes.push_back(rm);
        }
        reverse(right_maxes.begin(), right_maxes.end());

        int total = 0;
        for(int i = 0; i < height.size(); i++){
            total += max(0, min(left_maxes[i], right_maxes[i]) - height[i]);
        }
        return total;
    }
};